modules/code-deployment/src/main/java/org/apache/ignite/internal/deployunit/DeploymentManagerImpl.java

```diff
@@ -0,0 +1,435 @@
+        }
+
+        ByteArray key = new ByteArray(UNITS_PREFIX + id + ":" + version.render());
+
+        UnitMeta meta = new UnitMeta(id, version, deploymentUnit.name(), Collections.emptyList());
+
+        Operation put = put(key, UnitMetaSerializer.serialize(meta));
+
+        DeployUnitRequestBuilder builder = DeployUnitRequestImpl.builder();
+
+        try {
+            builder.unitContent(deploymentUnit.content().readAllBytes());
+        } catch (IOException e) {
+            LOG.error("Error to read deployment unit content", e);
+            return CompletableFuture.failedFuture(new DeploymentUnitReadException(e));
+        }
+        DeployUnitRequest request = builder
+                .unitName(deploymentUnit.name())
+                .id(id)
+                .version(version.render())
+                .build();
+
+        return metaStorage.invoke(notExists(key), put, Operations.noop())
+                .thenCompose(success -> {
+                    if (success) {
+                        return doDeploy(request);
+                    }
+                    LOG.error("Failed to deploy meta of unit " + id + ":" + version);
+                    return CompletableFuture.failedFuture(new DeployUnitWriteMetaException());
+                })
+                .thenApply(completed -> {
+                    if (completed) {
+                        startDeployAsyncToCmg(request);
+                    }
+                    return completed;
+                });
+    }
+
+    private void startDeployAsyncToCmg(DeployUnitRequest request) {
+        cmgManager.cmgNodes()
+                .thenAccept(nodes -> {
+                    for (String node : nodes) {
+                        ClusterNode clusterNode = clusterService.topologyService().getByConsistentId(node);
+                        if (clusterNode != null) {
+                            inFlightFutures.registerFuture(requestDeploy(clusterNode, request));
+                        }
+                    }
+                });
+    }
+
+    private CompletableFuture<Boolean> requestDeploy(ClusterNode clusterNode, DeployUnitRequest request) {
+        return clusterService.messagingService()
+                .invoke(clusterNode, request, Long.MAX_VALUE)
+                .thenCompose(message -> {
+                    Throwable error = ((DeployUnitResponse) message).error();
+                    if (error != null) {
+                        LOG.error("Failed to deploy unit " + request.id() + ":" + request.version()
+                                + " to node " + clusterNode, error);
+                        return CompletableFuture.failedFuture(error);
+                    }
+                    return CompletableFuture.completedFuture(true);
+                });
+    }
+
+    @Override
+    public CompletableFuture<Void> undeployAsync(String id, Version version) {
+        if (id == null || id.isBlank() || version == null) {
```

Comment:
Same as above: use `IllegalArgumentException`, separate check for every argument.